src/App.tsx

```diff
@@ -1,24 +1,178 @@
+      }, 500));
+  };
+
+  const addTodo = (title: string) => {
+    setIsLoading(true);
+    setProcessingTodoIds(prevIds => [...prevIds, 0]);
+    setTempTodo({
+      id: 0,
+      title,
+      completed: false,
+      userId: USER_ID,
+    });
+    todoService.addTodo({
+      title,
+      completed: false,
+      userId: USER_ID,
+    })
+      .then(newTodo => {
+        setTodoTitle('');
+        setTimeout(() => {
+          setTodos(currentTodos => {
+            return [...currentTodos, newTodo];
+          });
+          setTempTodo(null);
+        }, 500);
+      })
+      .catch(() => setErrorType(Errors.Add))
+      .finally(() => {
+        setIsLoading(false);
+        setProcessingTodoIds(prevIds => [...prevIds]
+          .filter(prevId => prevId !== 0));
+      });
+  };
+
+  const updateTodo = (updatedTodo: Todo) => {
+    setProcessingTodoIds(currentTodos => [...currentTodos, updatedTodo.id]);
+
+    todoService.updateTodo(updatedTodo)
+      .then(() => setTodos(prev => (
+        prev.map(prevTodo => (
+          prevTodo.id === updatedTodo.id
+            ? updatedTodo
+            : prevTodo
+        ))
+      )))
+      .catch(() => setErrorType(Errors.Update))
+      .finally(() => {
+        setProcessingTodoIds(currentTodos => currentTodos
+          .filter(todoId => updatedTodo.id !== todoId));
+      });
+  };
+
+  const toggleAll = async () => {
+    const isAllCompleted = todos.every(todo => todo.completed);
+
+    const todosToUpdate = todos.filter(todo => (isAllCompleted
+      ? todo.completed
+      : !todo.completed
+    ));
+
+    await Promise.all(todosToUpdate.map(todo => (
+      updateTodo({
+        ...todo,
+        completed: !isAllCompleted,
+      })
+    )));
+  };
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const isThereCompleted = todos.some(todo => todo.completed);
+
+  const filteredTodos = preparedTodos(todos, filterStatus);
+
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">React Todo App - Add and Delete</a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <TodoHeader
+          todoTitle={todoTitle}
+          setTodoTitle={setTodoTitle}
+          todos={todos}
+          isLoading={isLoading}
+          onAddTodo={addTodo}
+          setError={setErrorType}
+          onToggleAll={toggleAll}
+        />
+
+        <TodoList
+          todos={filteredTodos}
+          tempTodo={tempTodo}
+          onDeleteTodo={handleDelete}
+          processingTodoIds={processingTodoIds}
+          onUpdateTodo={updateTodo}
+        />
+
+        {todos.length !== 0 && (
```

Comment:
```suggestion
        {todos.length && (
```

Reply:
there is 0 appearing on the page in that variant